File: compiler/nodes/patterns/number.go
package patterns

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type NumberPatternNode struct {
	Value string
	Facts *database.Facts

	matching database.Node
}

func (node *NumberPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseNumberPattern(parser *syntax.Parser) (*NumberPatternNode, *syntax.Error) {
	span := parser.Spanned()

	value, err := parser.Token("Number")
	if err != nil {
		return nil, err
	}

	return &NumberPatternNode{
		Value: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *NumberPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	numberType := &types.NamedTypeNode{
		Name:       "Number",
		Parameters: []database.Node{},
		Facts:      database.NewFacts(database.GetSpanFact(node)),
	}

	visitor.Visit(numberType)

	visitor.Constraint(typecheck.GroupConstraint(node, numberType))
}

func (node *NumberPatternNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	c.WriteString(span, " && (")
	c.WriteNode(span, node.matching)
	c.WriteString(span, "=== ")
	c.WriteString(span, node.Value)
	c.WriteString(span, ")")

	return nil
}

File: compiler/src/nodes/patterns/number.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{HiddenNode, Node, NodeRef},
    nodes::{Matching, NamedTypeNode, visit_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{MatchPathSegment, Visit, Visitor},
};

#[derive(Debug)]
pub struct NumberPatternNode {
    pub value: String,
}

impl Node for NumberPatternNode {}

pub fn parse_number_pattern(parser: &mut Parser<'_>) -> Result<NumberPatternNode, ParseError> {
    let value = parser.token(TokenKind::Number)?;

    Ok(NumberPatternNode { value })
}

impl Visit for NumberPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, Some(MatchPathSegment::NoMatch));

        let span = visitor.span(node);
        let number_type = visitor.node(
            span,
            HiddenNode::new(NamedTypeNode {
                name: String::from("Number"),
                parameters: Vec::new(),
            }),
        );

        visitor.visit(&number_type);
        visitor.constraint(GroupConstraint::new(node.clone(), number_type));
    }
}

impl Codegen for NumberPatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Matching(matching) = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        ctx.condition(ir::Condition::EqualToNumber {
            input: matching,
            value: self.value.clone(),
        });

        Ok(())
    }
}
